### ISSUE TITLE
Use UUPS Upgrade to handle circular dependencies during initial deployment

### ISSUE THREAD

**Issue by immortal-tofu** [zama-ai/fhevm-backend#173]
Today, we need a fresh wallet with a 0 nonce to deploy contracts and we are using precomputed address. This workflow could be more flexible as follow:

1. Deploy all UUPS proxies with an empty implementation of the core contracts.
2. Get addresses of the deployed proxies.
3. Upgrade all proxies with a new implementation using correct addresses.

This method would cost more gas but is more flexible to handle circular dependencies. Each upgrade/deploy for each core contract should have their own task. Also, avoiding `.env` could be convenient (ie: deploying contract in a CI without context)

**Comment by jatZama** [comment]
> Would only cost more gas on deploy right? Since the contract addresses would be baked into the deployed contract code via an ordinary constructor on the implementation contracts. So I'd consider that pretty minor in grand scheme of things.

Exactly, the tradeoff of paying slightly more gas in order to have an easier deployment process is acceptable imho.

**Comment by bcm-at-zama** [comment]
Mentioned explicitly in the requests by Inco

**Comment by jatZama** [comment]
This was done already in #233 
So the following OKR https://github.com/zama-ai/planning-blockchain/issues/407 is considered done now.
